src/ksm/client/discovery.go

```diff
@@ -34,11 +35,42 @@ const (
 	headlessServiceClusterIP = "None"
 )
 
-type lookupSRVFunc func(service, proto, name string) (cname string, addrs []*net.SRV, err error)
+type LookupSRVFunc func(service, proto, name string) (cname string, addrs []*net.SRV, err error)
+
+// DiscovererConfig holds parameters for creating discoverer.
+type DiscovererConfig struct {
+	LookupSRV         LookupSRVFunc
+	APIClient         client.Kubernetes
+	Logger            *logrus.Logger
+	OverridenEndpoint string
+}
+
+func (dc DiscovererConfig) New() (client.Discoverer, error) {
```

Comment:
Having a config returning an instance of the object is an interesting concept. Normally I'm more used to do `func NewDiscovery(dc DiscovererConfig) (client.Discoverer, error)`.

Not that I'm against it or anything, but curious why you decided to do it this way :)

Reply:
Hmm, I usually do this to avoid passing the parameters to function, as those tend to grow. It's also sometimes useful to make struct the _context_ of function call, while function arguments can act more as a "runtime" dependencies, like context or logger, which often cannot be serialized.

However, this approach has a drawback, as it's not explicit what will be returned from `DiscovererConfig.New()`. One may guess, but `NewDiscovery(dc DiscovererConfig` is more explicit.

I wonder about `DiscovererConfig.NewDiscoverer()` now :thinking:

Reply:
Alternatively, just `DiscovererConfig.Discoverer()` could be a way as well.

But I do think that `NewDiscoverer(dc DiscovererConfig)` is perhaps what go developers are more used to see.

Reply:
Changed to `NewDiscoverer(dc DiscovererConfig)` then.